### ISSUE TITLE
what is the main speed up trick for nanoGPT?

### ISSUE THREAD

**Issue by brando90** [karpathy/nanoGPT#58]
question

curiosity, does someone know what the main trick to having nano GPT2 train so quickly? (from the 1 week I was used to to 1day it seems) https://github.com/karpathy/nanoGPT After a discussion it seems its only a smaller batch size...then it seems to achieve the same val loss quicker due to this. Is this the main trick really? Also doesn't a smaller batch size give us more uncertainty on the loss function? How do we know the two model truly perform the same? e.g. see confidence intervals on means depends on sqrt of N.

**Comment by karpathy** [comment]
Just to clarify, we're talking about GPT-2 124M here?

**Comment by brando90** [comment]
on a related note:

> Related to the previous, I've also been unsure about the role of the batch size in training LLMs compared to traditional deep learning. In traditional deep learning when we used epochs to train, a model the larger the batch size the quicker we could go through an epoch -- so the advice I received was to make the batch size large. Intuitively, the larger the batch size the more data the model sees per iteration. But mathematically this only really improves the variance of the gradient -- which isn't immediately obvious is what we want (I've done experiments and seen papers where noisy gradients lead to better models). It is clear to the that the larger the context size the better (for everything, but for the sake of this conv it's better for training) -- whenever possible.  But context size is totally different from batch size. So my question is, how does distributed training, especially at the node level help at all if batch size isn't really the helping factor (which might be a wrong assumption)? So the only role for distributed training I see is if the model is to large to fit in 1 node -- since I'm arguing there is no point to make the batch size too large (I'd guess 64-32 is fine due to the CLT).


How does the above related to nano gpt2?
